pageserver/src/bin/pageserver.rs

```diff
@@ -486,23 +509,49 @@ fn start_pageserver(conf: &'static PageServerConf) -> Result<()> {
             endpoint::serve_thread_main(router, http_listener)
         })?;
 
-    join_handles.push(http_endpoint_thread);
-
     // Spawn a thread to listen for connections. It will spawn further threads
     // for each connection.
     let page_service_thread = thread::Builder::new()
-        .name("Page Service thread".into())
+        .name("1 Page Service thread".into())
         .spawn(move || {
             page_service::thread_main(conf, auth, pageserver_listener, conf.auth_type)
         })?;
 
-    join_handles.push(page_service_thread);
-
-    for handle in join_handles.into_iter() {
-        handle
-            .join()
-            .expect("thread panicked")
-            .expect("thread exited with an error")
+    for info in &mut signals {
+        info!("Received a signal {:?}", info);
+        match info.signal {
+            SIGQUIT => {
+                info!("Got SIGQUIT. Terminate pageserver in immediate shutdown mode");
+                // TODO Do we need any actions here?
+                exit(1);
+            }
+            SIGTERM => {
+                info!("Got SIGINT/SIGTERM. Terminate gracefully in fast shutdown mode");
+                // Terminate postgres backends
+                postgres_backend::set_pgbackend_shutdown_requested();
+                // Stop all tenants and flush their data
+                tenant_mgr::shutdown_all_tenants()?;
+                // Wait for pageservice thread to complete the job
+                page_service_thread
+                    .join()
+                    .expect("thread panicked")
+                    .expect("thread exited with an error");
+
+                // Shutdown http router
```

Comment:
Does it make sense to shutdown http later, than the pageserver itself? 
What would http requests see when the server is shut already?

Also, would it make sense to try and shut down all threads, ignoring the failures? So, instead of panicking with `.expect("thread exited with an error");` and similar, we would rather log that sad event and go on with shutting down the rest?

Reply:
> Does it make sense to shutdown http later, than the pageserver itself? What would http requests see when the server is shut already?

I rearranged the code to shut down http the last, because it is now used to check the status while pageserver is shutting down. To handle this properly, we need to introduce pageserver states i.e. `starting up`, `running`, `shutting down`, and return them in `check_status()`.

> Also, would it make sense to try and shut down all threads, ignoring the failures? So, instead of panicking with `.expect("thread exited with an error");` and similar, we would rather log that sad event and go on with shutting down the rest?

Hmm, In my understanding if some thread is failed to shutdown clearly, we should consider shutdown failed. 
Do you think it is too strict and we should only care about the checkpoint success?